lib/charliecloud.py

```diff
@@ -701,6 +702,26 @@ def log(msg, hint, trace, color, prefix, end="\n"):
    if (color is not None):
       color_reset(log_fp)
 
+def monkey_write_streams():
+   """Monkey patch to replace problematic characters in stdout and stderr
+      streams when running Python 3.6. (see #1629)."""
+   def monkey_write_insert(f):
+      write_orig = f.write
+      def write_monkey(text):
+         text = text.replace("“", "\"").replace("”", "\"").replace("’", "'")
+         write_orig(text)
+      f.write = write_monkey
+   # Try to encode test string of problematic characters. If unsuccessful,
+   # monkey patch them out.
+   for stream in sys.stdout, sys.stderr:
+      for encoding in stream.encoding, locale.getpreferredencoding().lower(), "ASCII":
+         try:
+               "“”’".encode(encoding=encoding)
+         except UnicodeEncodeError:
+            monkey_write_insert(sys.stdout)
+            monkey_write_insert(sys.stderr)
+            break
```

Comment:
There's some subtle errors in the loop handling here, I think. I'd recommend something like (untested):

```suggestion
   for stream in sys.stdout, sys.stderr:
      for encoding in stream.encoding, locale.getpreferredencoding(), "ASCII":
         if (encoding is not None):
            try:
               "“”’".encode(encoding=encoding)
            except UnicodeEncodeError:
               monkey_write_insert(stream)
               break
```